### PR TITLE
Create new validation pattern

### PR DESCRIPTION
**Pull request by StephenGill**
This PR creates the new validation pattern and makes minor changes to two linked pages

### CODE REVIEW

File: src/patterns/validation/index.md.njk [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
++ [‘page not found’ pages](/../patterns/page-not-found-pages/)
++ [‘service unavailable’ pages](/../patterns/service-unavailable-pages/)
+
+## How it works
+
+Validation should refuse to accept:
+
++ information that can’t be correct
++ information that’s too ambiguous for you to use
++ missing information, it the information is required
```
Comment (joelanman) [inline src/patterns/validation/index.md.njk]: typo - 'it'

**Comment by 36degrees** [comment]
The tests are failing because there are 8 broken links:

```
patterns/validation/index.html >>> href: "/../patterns/problem-with-the-service-pages/", text: "‘there is a problem with the service’ pages"
patterns/validation/index.html >>> href: "/../patterns/page-not-found-pages/", text: "‘page not found’ pages"
patterns/validation/index.html >>> href: "/../patterns/service-unavailable-pages/", text: "‘service unavailable’ pages"
patterns/validation/index.html >>> href: "../components/error-summary/", text: "error summary"
patterns/validation/index.html >>> href: "../components/error-messages/", text: "error messages"
patterns/validation/index.html >>> href: "../error-message#be-clear-and-concise", text: "writing good error messages"
components/error-summary/index.html >>> href: "/../patterns/validation/", text: "validation pattern"
components/error-message/index.html >>> href: "/../patterns/validation/", text: "validation pattern"
```

File: src/patterns/validation/index.md.njk [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+For example, if you’re asking for someone’s date of birth you shouldn’t accept ‘13’ in the month field.
+
+### Telling the user about validation errors
+
+If the user's answers fail validation:
+
++ show them the page again, with the form fields as the user filled them in
++ add ‘Error: ’ to the beginning of the page `<title>` so screen readers read it out as soon as possible
++ show an [error summary](/components/error-summary) at the top of the page
++ show [error messages](/components/error-messages) next to fields with errors
```
Comment (36degrees) [inline src/patterns/validation/index.md.njk]: This link is broken – it should be:

```suggestion
+ show [error messages](/components/error-message/) next to fields with errors
```

File: src/patterns/validation/index.md.njk [hunk 3]
```diff
@@ -8,83 +8,87 @@ backlog_issue_id:
...
+
+### When to tell the user about validation errors
+
+Do not validate when the user moves away from a field. Wait until they try to move to the next part of the service - usually by clicking the ‘continue’ or ‘submit’ button at the bottom of the page.
 
-Read guidance on [writing good error messages](/components/error-message#be-clear-and-concise).
+To start with, avoid validating the information in a field while the user is still typing it in. This sort of validation can cause problems - especially for users who type more slowly.
 
-Show the error messages when they click the ‘continue’ or ‘submit’ button at the bottom of the page. Unless your user research tells you there’s a good reason to do so, avoid showing the error message:
+Only add this sort of validation if your user research shows that, on balance, it solves more problems for users than it causes. For example, the [character count component](/components/character-count/) shows users an error message when they go over the character limit. Because it’s important that users do not spend time and effort writing out a response that turns out to be too long.
```
Comment (joelanman) [inline src/patterns/validation/index.md.njk]: this is great, nice work

**Comment by StephenGill** [comment]
> I'd also challenge the use of "typing" over "entering" the information. I don't think this guidance should just apply to keyboard entry. There are a range of other interfaces this could also apply to, for example sliders or calendars.

Thanks - I've changed it to say 'don't validate while the user is still entering information'

**Comment by StephenGill** [comment]
> The guidance for this component is an excellent example of how to implement validation during typing, what to consider and when not to do it. It might be worth either saying that here or surfacing some of that content instead.

Agree - though on the new validation page, I think it's more about giving a high level example of a legitimate need for validation while you type, rather than details of how to implement it. And I think we've done that ok?

**Comment by StephenGill** [comment]
> Being a bit picky but not sure what "To start with" adds here.
> 
> I also think we might need another heading, possibly a sub-heading below "When to tell the user about validation errors" for validating during input.

It's a matter of opinion, but I think the subheads look ok as they are

**Comment by StephenGill** [comment]
Think this is ready for merging - @joelanman would you mind giving it a final review?

File: src/patterns/validation/index.md.njk [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
++ the visual style, placement and content of HTML5 error messages can’t be made consistent with the GOV.UK Design System
++ we know that the GOV.UK Design System error message and error summary components are accessible
+
+To turn off HTML5 validation:
+
++ add ‘[novalidate](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-novalidate)’ to your form tags
++ do not add ‘required’ to your input fields
+
+## Help improve this page
```
Comment (fofr) [inline src/patterns/validation/index.md.njk]: From the preview it looks like there's two "Help improve this page" sections:
https://deploy-preview-1164--govuk-design-system-preview.netlify.com/patterns/validation/

Screen Shot 2020-03-04 at 09 18 31

Comment (StephenGill) [reply src/patterns/validation/index.md.njk]: Thanks - I forgot that it adds a 'Help improve this page' paragraph automatically. Should be fixed now.

File: src/patterns/validation/index.md.njk [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+HTML5 validation is a type of client side validation built into browsers. Do not use it because:
+
++ the visual style, placement and content of HTML5 error messages can’t be made consistent with the GOV.UK Design System
++ we know that the GOV.UK Design System error message and error summary components are accessible
+
+To turn off HTML5 validation:
+
++ add ‘[novalidate](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-novalidate)’ to your form tags
++ do not add ‘required’ to your input fields
```
Comment (paulwaitehomeoffice) [inline src/patterns/validation/index.md.njk]: Do users have to refrain from using the `required` attribute? `novalidate` disables in-browser validation triggered by it, as far as I can see, and I'm fairly sure some screen readers will say "required" when reading form fields with this attribute, which might (or, admittedly, might not) be useful for screen reader users.

File: src/components/error-message/index.md.njk [hunk 6]
```diff
@@ -1,6 +1,6 @@
 ---
 title: Error message
-description: Use an error message when there is a validation error. Explain what went wrong and how to fix it
+description: When there's a validation error, use an error message to explain what went wrong and how to fix it
```
Comment (edwardhorsford) [inline src/components/error-message/index.md.njk]:
```suggestion
description: When there’s a validation error, use an error message to explain what went wrong and how to fix it
```

File: src/patterns/validation/index.md.njk [hunk 7]
```diff
@@ -0,0 +1,92 @@
+---
+title: Recover from errors (validation)
+description: Check the answers users give to make sure they’re valid - and if there’s an error, tell them what's wrong and how to fix it
```
Comment (edwardhorsford) [inline src/patterns/validation/index.md.njk]:
```suggestion
description: Check the answers users give to make sure they’re valid - and if there’s an error, tell them what’s wrong and how to fix it
```

File: src/patterns/validation/index.md.njk [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+section: Patterns
+theme: Help users to…
+aliases: 
+backlog_issue_id: 
+layout: layout-pane.njk
+---
+
+Check the information the user gives you to make sure it’s valid. If there's a validation error, tell the user what's gone wrong and how to fix it. Turn off HTML5 validation.
```
Comment (edwardhorsford) [inline src/patterns/validation/index.md.njk]:
```suggestion
Check the information the user gives you to make sure it’s valid. If there’s a validation error, tell the user what's gone wrong and how to fix it. Turn off HTML5 validation.
```

File: src/patterns/validation/index.md.njk [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+## When to use this pattern
+
+Validation should be a last resort. Minimise your chances of needing to show an error message by:
```
Comment (edwardhorsford) [inline src/patterns/validation/index.md.njk]: > Validation should always be a last resort

I find this odd phrasing. Services should try to do the hard work so it's less common - but validation is expected on the majority of service pages. Of the services I previously assessed, validation was often forgotten about / not given enough attention.

File: src/patterns/validation/index.md.njk [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
++ [making sure your questions are well designed](https://www.gov.uk/service-manual/design/designing-good-questions)
++ accepting information in different formats, as long as it’s not ambiguous (for example, you should accept postcodes with or without spaces - or names with non-alphabetical characters like accents)
```
Comment (edwardhorsford) [inline src/patterns/validation/index.md.njk]: > alphabetical 

I think these are diacritics. Aren't letters like èáö still alphabetical?

Comment (edwardhorsford) [reply src/patterns/validation/index.md.njk]: Could include some examples of diacritic characters.

File: src/patterns/validation/index.md.njk [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
++ information that can’t be correct
++ information that’s too ambiguous for you to use
++ missing information, if the information is required
+
+For example, if you’re asking for someone’s date of birth you shouldn’t accept ‘13’ in the month field.
+
+### How to tell the user about validation errors
+
+If the user's answers fail validation:
```
Comment (edwardhorsford) [inline src/patterns/validation/index.md.njk]:
```suggestion
If the user’s answers fail validation:
```

File: src/patterns/validation/index.md.njk [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+## Turn off HTML5 validation
+
+HTML5 validation is a type of client side validation built into browsers. Do not use it because:
+
++ the visual style, placement and content of HTML5 error messages can’t be made consistent with the GOV.UK Design System
++ we know that the GOV.UK Design System error message and error summary components are accessible
+
+To turn off HTML5 validation:
+
++ add ‘[novalidate](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-novalidate)’ to your form tags
```
Comment (edwardhorsford) [inline src/patterns/validation/index.md.njk]:
```suggestion
+ add ‘[novalidate](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-novalidate)’ to your `form` tags
```

File: src/patterns/validation/index.md.njk (hunk 5 above)
Comment (edwardhorsford) [inline src/patterns/validation/index.md.njk]: I think this is unrelated to HTML5 validation.

File: src/patterns/validation/index.md.njk [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
++ server side validation
++ client side validation (in the browser, using either JavaScript or HTML)
+
+You’ll always need to carry out server side validation, because there’s no guarantee that client side validation will work in all circumstances. For example users can bypass client side validation, or [JavaScript can fail to load](https://www.gov.uk/service-manual/technology/using-progressive-enhancement).
+
+Avoid adding client side validation unless you can establish that there’s a need for it. For example, because user research shows it’s helpful to users if you validate the information they’re providing as they type.
+
+If you do use client side validation, make sure your client side and server side validation rules do not conflict.
+
```
Comment (edwardhorsford) [inline src/patterns/validation/index.md.njk]: Is it worth mentioning some of the common pitfalls of client side validation?

For instance:
* Distracting / jarring / trigger to easily
* Reflows the page

File: src/patterns/validation/index.md.njk [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+backlog_issue_id: 
+layout: layout-pane.njk
+---
+
+Check the information the user gives you to make sure it’s valid. If there's a validation error, tell the user what's gone wrong and how to fix it. Turn off HTML5 validation.
+
+## When to use this pattern
+
+Validation should be a last resort. Minimise your chances of needing to show an error message by:
+
```
Comment (edwardhorsford) [inline src/patterns/validation/index.md.njk]: Could add something about reasonably normalising input, for example by:
* Stripping leading and trailing whitespace
* Converting / stripping characters you understand but don't need - eg brackets in telephone numbers.
* Converting case on things like postcode and reference numbers.

File: src/components/error-message/index.md.njk (hunk 6 above)
Comment (36degrees) [reply src/components/error-message/index.md.njk]: The Markdown parser we're using takes care of quotes automatically, converting them to the appropriate character.

File: src/patterns/validation/index.md.njk (hunk 5 above)
Comment (joelanman) [reply src/patterns/validation/index.md.njk]: Thanks for this, we'll recommend not using `required` for now as that's consistent with out approach in general (to only highlight optional fields). However we've added a line to the Research section - to check whether this is the right approach for screen reader users.

File: src/patterns/validation/index.md.njk (hunk 9 above)
Comment (joelanman) [reply src/patterns/validation/index.md.njk]: Thank yes, we've updated the language here.

File: src/patterns/validation/index.md.njk (hunk 13 above)
Comment (joelanman) [reply src/patterns/validation/index.md.njk]: We don't want to get too much into the technical aspects of client side validation, as we're not recommending it. If teams find a need to add it, they will have to look into accessibility, reliability etc for themselves. If common user needs are found, we could document tried and tested approaches.

File: src/patterns/validation/index.md.njk (hunk 14 above)
Comment (joelanman) [reply src/patterns/validation/index.md.njk]: Normalising might be something to cover in other guidance - for now we just want to focus on validation. Another related area we might get into later is sanitation - making input safe to store and display, but yes we wanted to keep the scope just on validation here.

File: src/patterns/validation/index.md.njk [hunk 15]
```diff
@@ -0,0 +1,93 @@
+---
+title: Recover from errors (validation)
```
Comment (timpaul) [inline src/patterns/validation/index.md.njk]: Could we tweak the title to remove the parentheses please? Ta

File: src/components/error-summary/index.md.njk [hunk 16]
```diff
@@ -21,7 +21,14 @@ Always show an error summary when there is a validation error, even if there’s
 
 ## How it works
 
-As well as showing an error summary, follow the [validation pattern](/patterns/validation) - for example, by adding ‘Error: ’ to the beginning of the page `<title>`.
+You must:
+
++ move keyboard focus to the error summary
++ include the heading ‘There is a problem’
++ link to each of the answers that have validation errors
++ make sure the error messages in the error summary are worded the same as those which appear next to the inputs with errors
```
Comment (edwardhorsford) [inline src/components/error-summary/index.md.njk]: In most cases I reckon this is correct - but we had cases when working on Passports where it made sense for the error next to the field to be different from the error in the summary list. I'll try to find an example.

File: src/patterns/validation/index.md.njk (hunk 10 above)
Comment (StephenGill) [reply src/patterns/validation/index.md.njk]: Thanks @edwardhorsford  - have clarified that validation should accept diacritics as well as non-alphabetical characters

**Comment by StephenGill** [comment]
Should be ready for merging now. I've:

- removed the parentheses from the h1
- clarified that name validation should accept diacritics as well an non-alphabetical characters, as per, @edwardhorsford's suggestion
- added a couple of points back into the checklist about how and when to display the error summary component
- made some minor tweaks for clarity

@joelanman @timpaul

**Comment by joelanman** [comment]
Thanks @StephenGill, looks good to me

**Comment by 36degrees** [comment]
Can we rebase this and try and squash down from 20 commits before merging please.
